fv3core/decorators.py

```diff
@@ -149,7 +149,96 @@ def __iter__(self):
         return self.cache.__iter__()
 
 
-class FV3StencilObject:
+class StencilWrapper:
+    """Wrapped GT4Py stencil object."""
+
+    def __init__(
+        self,
+        func: Callable[..., None],
+        origin: Optional[Index3D] = None,
+        domain: Optional[Index3D] = None,
+        **kwargs,
+    ):
+        self.func = func
+        self.origin = origin
+        self.domain = domain
+
+        if "format_source" not in kwargs:
+            kwargs["format_source"] = global_config.get_format_source()
+
+        self.stencil_object = gtscript.stencil(
+            definition=self.func,
+            backend=global_config.get_backend(),
+            rebuild=global_config.get_rebuild(),
+            **kwargs,
+        )
+
+    def __call__(self, *args, **kwargs) -> None:
+        if self.origin:
+            assert "origin" not in kwargs, "Cannot override origin provided at init"
+            kwargs["origin"] = self.origin
+        kwargs["origin"] = self._get_field_origins(*args, **kwargs)
+
+        if self.domain:
+            assert "domain" not in kwargs, "Cannot override domain provided at init"
+            kwargs["domain"] = self.domain
+        else:
+            assert "domain" in kwargs, "No domain provided at call time"
+
+        if global_config.get_validate_args():
+            self.stencil_object(*args, **kwargs, validate_args=True)
+        else:
+            kwargs = self._process_kwargs(*args, **kwargs)
+            self.stencil_object.run(**kwargs, exec_info=None)
+
+    def _process_kwargs(self, *args, **kwargs):
+        """Process keyword args so they can be passed directly to stencil_object.run."""
+
+        for keyword in ("origin", "domain"):
```

Comment:
pop with None as the secondary value lets you delete if present if that is easier?

Reply:
Good idea!